reference/src/layout/arrays.md

```diff
@@ -0,0 +1,76 @@
+> foo(T x[N])`, decays to a pointer. That is, these functions do not take arrays
+> as an arguments, they take a pointer to the first element of the array
+> instead. Array types are therefore _improper C types_ (not C FFI safe) in Rust
+> foreign function declarations, e.g., `extern { fn foo(x: [T; N]) -> [U; M];
+> }`. Pointers to arrays are fine: `extern { fn foo(x: *const [T; N]) -> *const
+> [U; M]; }`.
+
+The _stride_ of the array is the distance between a pair of consecutive values
+within the array, and it is constant for all element pairs. It is computed as
+the _size_ of the element type rounded up to the next multiple of the
+_alignment_ of the element type.
+
+When the element _size_ is a multiple of the element's _alignment_, then `stride
+== size`, and the elements are laid out contiguously in memory, e.g., `[u8; 4]`.
+In this case, the _size_ of the array can be computed as `size_of::<T>() * N`,
+and a pointer to the `i`-th element of the array can be obtained by offsetting a
+pointer to the first element of the array by `i`[^1].
+
+> **Note:** In the current Rust implementation, _size_ is always a multiple of
+> the element's _alignment_, and therefore `stride == size` always holds. This
+> is, however, not guaranteed by the [layout of structs and tuples].
+
+[^1]: When `stride > size` the pointer needs to be advanced by the array
+    _stride_ instead of by the element _size_.
+
+[layout of structs and tuples]: ./structs-and-tuples.md
+
+The [layout of Vector types][Vector] [^2] requires the _size_ and _alignment_ of
+the [Vector] elements to match. That is, types with [Vector] layout are layout
+compatible with arrays having the same element type and the same number of
+elements as the [Vector].
+
+[^2]: The [Vector] layout is the layout of `repr(simd)` types like [`__m128`].
+
+[Vector]: vectors.md
+[`__m128`]: https://doc.rust-lang.org/core/arch/x86_64/struct.__m128.html
+
+## Unresolved questions
+
+### Guaranteeing `stride == size` ?
+
+Currently, the [layout of structs and tuples] does not guarantee that the
+element _size_ is a multiple of its _alignment_. For example, consider:
+
+```rust,ignore
+struct A(u16, u8);
+type B = [A; 4];
+```
+
+In the current Rust implementation, `A` has an alignment and a size of `4`, and
+`B` has a size of `16`, such that `B` contains four `A`s that are contiguously
+laid in memory. 
+
+However, a future Rust implementation could implement a layout optimization that
+reduces the size of `A` to `3`. For the elements of `B` to be properly aligned,
+`B` would need to choose a `stride == 4`, resulting in a `stride > size`.
+
+Guaranteeing `stride >= size` is forward-compatible with such
+layout-optimization proposals:
+  
+  * [rust-lang/rfcs/1397: Spearate size and stride for types](https://github.com/rust-lang/rfcs/issues/1397)
+  * [rust-lang/rust/17027: Collapse trailing padding](https://github.com/rust-lang/rust/issues/17027)
```

Comment:
Shouldn't slices be mentioned somewhere?

Reply:
I don't know. What should we say about them?

Reply:
Well it seems to me that the layout of arrays and the layout of a slice are closely related -- they must have the same stride etc. Otherwise unsizing would be wrong.

Reply:
Does that belong here or should there be a "Layout of slices" document ?

I can rename this document to "Arrays and Slices", and add a sub-section about the layout of slices, where the text could say:

"The layout of a slice `[T]` with `N` elements is the same as that of a `[T; N]` array. "

But I am not sure if that's enough.

Reply:
It seems like a good start, at the least.

Reply:
So I've added a sub-section for it.